Source/CarthageKit/NewResolver.swift

```diff
@@ -0,0 +1,555 @@
+		guard hasNext else { return }
+
+		// 'skip' any permutations as defined by 'startingIndex' by setting all subsequent values to their max. We don't count this as an 'incremented' occurrence.
+		if let startingIndex = startingIndex {
+			for i in (startingIndex..<permutation.endIndex) {
+				permutation[i].skipRemaining()
+			}
+		}
+
+		// If we 'reset' for every dimension, we've hit the end
+		hasNext = false
+		for i in (permutation.startIndex..<permutation.endIndex).reversed() {
+			if permutation[i].increment() == .incremented {
+				hasNext = true
+				break
+			}
+		}
+	}
+}
+
+/// Helper struct to track a single axis of a permutation
+extension NodePermutations {
+	enum IncrementResult {
+		case incremented
+		case reset
+	}
+
+	struct Dimension {
+		let nodes: [DependencyNode]
+		var index: Array<DependencyNode>.Index
+
+		init(_ nodes: [DependencyNode]) {
+			self.nodes = nodes
+			self.index = nodes.startIndex
+		}
+
+		var node: DependencyNode {
+			return nodes[index]
+		}
+
+		mutating func skipRemaining() {
+			index = nodes.index(before: nodes.endIndex)
+		}
+
+		mutating func increment() -> IncrementResult {
+			index = nodes.index(after: index)
+			if index < nodes.endIndex {
+				return .incremented
+			} else {
+				index = nodes.startIndex
+				return .reset
+			}
+		}
+	}
+}
+
+/// A node in, or being considered for, an acyclic dependency graph.
+private final class DependencyNode {
+	/// The dependency that this node refers to.
+	let dependency: Dependency
+
+	/// The current requirements this dependency node was created from
+	var versionSpecifier: VersionSpecifier
+
+	/// The parent node where versionSpecifier was sourced from
+	var parent: DependencyNode?
+
+	/// The dependencies of this node.
+	var dependencies: Set<DependencyNode> = []
+
+	/// The version of the dependency that this node represents.
+	///
+	/// This version is merely "proposed" because it depends on the final
+	/// resolution of the graph, as well as whether any "better" graphs exist.
+	let proposedVersion: PinnedVersion
+
+	init(dependency: Dependency, proposedVersion: PinnedVersion, versionSpecifier: VersionSpecifier, parent: DependencyNode?) {
+		precondition(versionSpecifier.isSatisfied(by: proposedVersion))
+
+		self.dependency = dependency
+		self.proposedVersion = proposedVersion
+		self.versionSpecifier = versionSpecifier
+		self.parent = parent
+	}
+}
+
+extension DependencyNode: Comparable {
+	fileprivate static func < (_ lhs: DependencyNode, _ rhs: DependencyNode) -> Bool {
+		let leftSemantic = SemanticVersion.from(lhs.proposedVersion).value ?? SemanticVersion(major: 0, minor: 0, patch: 0)
+		let rightSemantic = SemanticVersion.from(rhs.proposedVersion).value ?? SemanticVersion(major: 0, minor: 0, patch: 0)
+
+		// Try higher versions first.
+		return leftSemantic > rightSemantic
+	}
+
+	fileprivate static func == (_ lhs: DependencyNode, _ rhs: DependencyNode) -> Bool {
+		return lhs.dependency == rhs.dependency
```

Comment:
I'm not comfortable with this definition of equality for `DependencyNode`. This should changed to true equality.

Reply:
Ok, this is the original comparison in the other resolver. I can make it true equality, but not sure how it'll affect the algorithm. I'll double check everything

Reply:
🙈